### PR TITLE
MF-500: Added 'Edit Patient Button' in Patient Actions Menu

### PR DESCRIPTION
**Pull request by vasharma05**
**What does this PR do?**

This PR is a solution to ticket [MF-500](https://issues.openmrs.org/browse/MF-500). This PR adds an `Edit patient Button` in the Patient's Actions Menu as is visible from the image added and redirects it to "patient/:patientUuid/edit" page to edit the details, using the already made Create/Update Patient Registration(Details).

This PR is a successor to https://github.com/openmrs/openmrs-esm-patient-registration/pull/80 , waiting to be merged on **Patient Registration** repo.

image

### CODE REVIEW

File: src/index.ts [hunk 1]
```diff
@@ -409,6 +410,20 @@ function setupOpenMRS() {
           }
         )
       },
+      {
+        id: "edit-patient-details-button",
+        slot: "patient-actions-slot",
+        load: getAsyncLifecycle(
+          () =>
+            import(
+              "./ui-components/custom-overflow-menu/custom-items/edit-patient-details"
```
Comment (FlorianRappl) [inline src/index.ts]: This whole thing seems to be best placed in patient registration.

After all the URL (and the respective parts) etc. are known only in patient registration.

Please put the component there.

**Review by FlorianRappl — CHANGES_REQUESTED** [review]
Wrong domain for the component.

`@openmrs/esm-patient-registration-app` seems right.

**Comment by brandones** [comment]
Right, the trouble is that the extension expected by the slot right now is this very particular `CustomOverflowMenuItem` thing. Ideally we'd be able to refactor it so that the extensions are just `<ConfigurableLink>` elements (or something similarly basic/universal) and the _extension slot_ does the work of putting them in `CustomOverflowMenuItem`.

This would be easier if we had support for links as a special class of extensions. I think there's a good case for it, at this point.

But it should be doable now, @vasharma05 , though maybe a bit tricky.

**Comment by vasharma05** [comment]
Hello @FlorianRappl and @brandones ! Sorry @FlorianRappl for being late, I just got notified. 

@brandones I got your point partially. But the issue is that, not all the menu item is a link. The `Start new visit` and `Add Past visit` calls a function and they are not a link. You can view the file in here. https://github.com/openmrs/openmrs-esm-patient-chart-widgets/blob/master/src/ui-components/custom-overflow-menu/custom-items/start-visit.tsx

**Comment by vasharma05** [comment]
Hello @brandones ! I gave a thought on what you said, to shift OverflowMenuItem into an extension. I feel like I can work on it and am now going to work to implement it.

**Comment by vasharma05** [comment]
Hello @FlorianRappl @brandones ! I have shifted the `CustomOverflowMenuItem` into Extension and it's all running good. Please review the new changes.
Thank You!

**Comment by vasharma05** [comment]
Solved merge conflicts

**Comment by brandones** [comment]
Feel free to mark as "ready for review" once you think a PR is in good shape.

**Comment by vasharma05** [comment]
Ohh, I didn't know about that. I'll take care of it next time.

**Comment by vasharma05** [comment]
@brandones about the changes that @FlorianRappl requested, I have shifted it to patient-registration using the Extension slot="custom-overflow-menu-item". 
Shall I pull the PR in there, can you review it?

**Comment by brandones** [comment]
Right, I think you missed the point about @FlorianRappl 's feedback. The idea is that a button that navigates to the registration page should be an extension in the registration module. I don't know why you've added these extra slots. The slot you already have, `patient-actions-slot`, is sufficient. You just need to create "edit-patient-button" extension in the registration module that attaches to that slot.

**Comment by vasharma05** [comment]
Yes, that is what I have done. 

I have added extension for `custom-overflow-menu-item` in here, so that I can use it in the `Edit Patient Details` button in the registration app, which is then put up in the `patient-actions-slot`.

File: src/index.ts [hunk 2]
```diff
@@ -406,6 +406,17 @@ function setupOpenMRS() {
           { featureName: "form-entry", moduleName }
         )
       },
+      {
+        id: "custom-overflow-menu-item",
+        slot: "custom-overflow-menu-item",
```
Comment (FlorianRappl) [inline src/index.ts:411]: Needs to end with `-slot`. Also it should describe the location, this one describes what / how the component looks like.

A good slot name would be something "more-patient-actions-slot".

File: src/ui-components/custom-overflow-menu/custom-items/add-past-visit.tsx [hunk 3]
```diff
@@ -1,16 +1,18 @@
...
 
 export default function AddPastVisitOverflowMenuItem() {
   const { t } = useTranslation();
   return (
-    <CustomOverflowMenuItem
-      itemText={t("Add Past Visit", "Add Past Visit")}
-      onClick={() =>
-        openVisitDashboard(`${t("visitDashboard", "Visit Dashboard")}`)
-      }
+    <ExtensionSlot
```
Comment (FlorianRappl) [inline src/ui-components/custom-overflow-menu/custom-items/add-past-visit.tsx:9]: Hm I am not 100% sure what you are trying to do here.

It seems that you want a very particular item to appear here. That's not what this should be.

**Review by FlorianRappl — CHANGES_REQUESTED** [review]
This looks like spaghetti.

What are you trying to achieve here?

**Comment by vasharma05** [comment]
Sorry @FlorianRappl ! I had a discussion with @brandones yesterday night, and forgot to mention it here. I have attached the screenshot in here. 
Now I am going to revert all the changes done and follow what @brandones suggested.
image

Should I create a new PR with the new changes and close this one or, should I push the new changes in this PR only?

**Comment by FlorianRappl** [comment]
You can reuse this PR.
Thanks @vasharma05

**Comment by vasharma05** [comment]
Sure! 
Thanks @FlorianRappl !

**Comment by FlorianRappl** [comment]
I did not see any changes here since the last message. Therefore I'll assume the PR is no longer needed.

In case I'm wrong just use your code and place it in the new monorepo.

**Comment by vasharma05** [comment]
Hello @FlorianRappl ! I shifted to work on the proposal and wasn't able to work on the PR. The work is pending and I am now working on it. 

What do you mean by monorepo?

**Comment by FlorianRappl** [comment]
> What do you mean by monorepo?

https://en.wikipedia.org/wiki/Monorepo

As you should know (since this was topic in the last two or three dev weeklys) the `openmrs-esm-patient-chart` repository is transformed to a mono repo containing the patient chart and all other patient-chart related microfrontends. The widgets are no longer and are instead broken up in individual microfrontends.

**Comment by vasharma05** [comment]
I was there when the discussion was going on to shift widgets into patient chart, but didn't know that they were shifted, since I was not active. 

Another fault from my side😅. Thank You @FlorianRappl once again!  I'll just have a look at them.

**Pull request closed without merging**